mono/utils/mono-threads-state-machine.c

```diff
@@ -206,37 +230,42 @@ MonoRequestSuspendResult
 mono_threads_transition_request_suspension (MonoThreadInfo *info)
 {
 	int raw_state, cur_state, suspend_count;
+	gboolean wont_safept;
 	g_assert (info != mono_thread_info_current ());
 
 retry_state_change:
-	UNWRAP_THREAD_STATE (raw_state, cur_state, suspend_count, info);
+	UNWRAP_THREAD_STATE (raw_state, cur_state, suspend_count, wont_safept, info);
 
 	switch (cur_state) {
 	case STATE_RUNNING: //Post an async suspend request
 		if (!(suspend_count == 0))
 			mono_fatal_with_history ("suspend_count = %d, but should be == 0", suspend_count);
-		if (mono_atomic_cas_i32 (&info->thread_state, build_thread_state (STATE_ASYNC_SUSPEND_REQUESTED, 1), raw_state) != raw_state)
+		if (mono_atomic_cas_i32 (&info->thread_state, build_thread_state (STATE_ASYNC_SUSPEND_REQUESTED, 1, wont_safept), raw_state) != raw_state)
 			goto retry_state_change;
-		trace_state_change ("SUSPEND_INIT_REQUESTED", info, raw_state, STATE_ASYNC_SUSPEND_REQUESTED, 1);
+		trace_state_change ("SUSPEND_INIT_REQUESTED", info, raw_state, STATE_ASYNC_SUSPEND_REQUESTED, wont_safept, 1);
 		return ReqSuspendInitSuspendRunning; //This is the first async suspend request against the target
 
-	case STATE_ASYNC_SUSPENDED:
-	case STATE_SELF_SUSPENDED:
 	case STATE_BLOCKING_SELF_SUSPENDED:
 	case STATE_BLOCKING_ASYNC_SUSPENDED:
+	case STATE_ASYNC_SUSPENDED:
+	case STATE_SELF_SUSPENDED:
+		if (wont_safept)
+			mono_fatal_with_history ("wont_safept = TRUE, but should be FALSE");
 		if (!(suspend_count > 0 && suspend_count < THREAD_SUSPEND_COUNT_MAX))
 			mono_fatal_with_history ("suspend_count = %d, but should be > 0 and < THREAD_SUSPEND_COUNT_MAX", suspend_count);
-		if (mono_atomic_cas_i32 (&info->thread_state, build_thread_state (cur_state, suspend_count + 1), raw_state) != raw_state)
+		if (mono_atomic_cas_i32 (&info->thread_state, build_thread_state (cur_state, suspend_count + 1, wont_safept), raw_state) != raw_state)
 			goto retry_state_change;
-		trace_state_change ("SUSPEND_INIT_REQUESTED", info, raw_state, cur_state, 1);
+		trace_state_change ("SUSPEND_INIT_REQUESTED", info, raw_state, cur_state, wont_safept, 1);
 		return ReqSuspendAlreadySuspended; //Thread is already suspended so we don't need to wait it to suspend
 
 	case STATE_BLOCKING:
 		if (!(suspend_count == 0))
 			mono_fatal_with_history ("suspend_count = %d, but should be == 0", suspend_count);
-		if (mono_atomic_cas_i32 (&info->thread_state, build_thread_state (STATE_BLOCKING_SUSPEND_REQUESTED, 1), raw_state) != raw_state)
+		if (wont_safept)
+			mono_fatal_with_history ("wont_safept = TRUE, but should be FALSE");
+		if (mono_atomic_cas_i32 (&info->thread_state, build_thread_state (STATE_BLOCKING_SUSPEND_REQUESTED, 1, wont_safept), raw_state) != raw_state)
 			goto retry_state_change;
-		trace_state_change ("SUSPEND_INIT_REQUESTED", info, raw_state, STATE_BLOCKING_SUSPEND_REQUESTED, 1);
+		trace_state_change ("SUSPEND_INIT_REQUESTED", info, raw_state, STATE_BLOCKING_SUSPEND_REQUESTED, wont_safept, 1);
```

Comment:
Contradictions really are ugly. `wont` is actually a word with a different meaning.

Reply:
i don't like `can_safepoint` because the bit applies to many states where safepointing just doesn't make sense:  `STATE_SELF_SUSPENDED | can_safepoint`, for example, just confuses me - it can't really safepoint - it's not even running.